services/leader/client.go

```diff
@@ -0,0 +1,220 @@
+	// import etcd's concurrency package in order to check the cause of the
+	// error.
+	ErrNoLeader = concurrency.ErrElectionNoLeader
+
+	// ErrCampaignInProgress is returned when a call to Campaign() is made while
+	// the caller is either already (a) campaigning or (b) the leader.
+	ErrCampaignInProgress = errors.New("campaign in progress")
+)
+
+// NB(mschalle): when an etcd leader failover occurs, all current leases have
+// their TTLs refreshed: https://github.com/coreos/etcd/issues/2660
+
+type client struct {
+	sync.RWMutex
+
+	ecl         *election.Client
+	opts        services.ElectionOptions
+	ctxCancel   context.CancelFunc
+	resignC     chan struct{}
+	campaigning uint32
+	closed      uint32
+}
+
+// newClient returns an instance of an client bound to a single election.
+func newClient(cli *clientv3.Client, opts Options, electionID string, ttl int) (*client, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	var sessionOpts []concurrency.SessionOption
+	if ttl != 0 {
+		sessionOpts = append(sessionOpts, concurrency.WithTTL(ttl))
+	}
+
+	pfx := electionPrefix(opts.ServiceID(), electionID)
+	ec, err := election.NewClient(cli, pfx, election.WithSessionOptions(sessionOpts...))
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		ecl:     ec,
+		opts:    opts.ElectionOpts(),
+		resignC: make(chan struct{}),
+	}, nil
+}
+
+func (c *client) campaign(opts services.CampaignOptions) (<-chan campaign.Status, error) {
+	if c.isClosed() {
+		return nil, errClientClosed
+	}
+
+	if !c.startCampaign() {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c.Lock()
+	c.ctxCancel = cancel
+	c.Unlock()
+
+	proposeVal := c.val(opts.LeaderValue())
+
+	// buffer 1 to not block initial follower update
+	sc := make(chan campaign.Status, 1)
+
+	sc <- campaign.NewStatus(campaign.Follower)
+
+	go func() {
+		defer func() {
+			close(sc)
+			cancel()
+			c.resetCampaign()
+		}()
+
+		// Campaign blocks until elected. Once we are elected, we get a channel
+		// that's closed if our session dies.
+		ch, err := c.ecl.Campaign(ctx, proposeVal)
+		if err != nil {
+			sc <- campaign.NewErrCampaignStatus(err)
+			return
+		}
+
+		sc <- campaign.NewStatus(campaign.Leader)
+		select {
+		case <-ch:
+			sc <- campaign.NewErrCampaignStatus(election.ErrSessionExpired)
+		case <-c.resignC:
+			sc <- campaign.NewStatus(campaign.Follower)
+		}
+	}()
+
+	return sc, nil
+}
+
+func (c *client) resign() error {
+	if c.isClosed() {
+		return errClientClosed
+	}
+
+	// if there's an active blocking call to Campaign() stop it
+	c.Lock()
+	if c.ctxCancel != nil {
+		c.ctxCancel()
+		c.ctxCancel = nil
+	}
+	c.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.opts.ResignTimeout())
+	defer cancel()
+	if err := c.ecl.Resign(ctx); err != nil {
+		return err
+	}
+
+	c.resetCampaign()
+
+	// if successfully resigned and there was a campaign in Leader state cancel
```

Comment:
Should we move this before `c.resetCampaign()`? Otherwise I can think of a scenario where this goroutine resets the campaign, then another goroutines starts a new campaign and becomes the leader, then this goroutine pushes a signal to the `resignC` channel, then the other goroutine reads it off the channel and becomes a follower even though it shouldn't.